compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -971,6 +979,66 @@ public void visit(BLangSimpleVariable varNode) {
         transferForkFlag(varNode);
     }
 
+    private void validateInferredArrays(BLangNode node, BLangExpression expr) {
+        switch (node.getKind()) {
+            case ARRAY_TYPE:
+                BLangArrayType arrayType = (BLangArrayType) node;
+                if (arrayType.inferredArrayValidateState == 1) {
+                    if (expr != null && expr.getKind() == NodeKind.LIST_CONSTRUCTOR_EXPR &&
+                            validateInitializer(((BLangArrayType) node).sizes,
+                            ((BLangArrayType) node).sizes.size() - 2, (BLangListConstructorExpr) expr)) {
+                        arrayType.inferredArrayValidateState = 2;
+                    }
+                }
+                return;
+            case INTERSECTION_TYPE_NODE:
+                BLangIntersectionTypeNode intersectionTypeNode = (BLangIntersectionTypeNode) node;
+                for (BLangType member : intersectionTypeNode.constituentTypeNodes) {
+                    validateInferredArrays(member, expr);
+                }
+                return;
+            case UNION_TYPE_NODE:
+                BLangUnionTypeNode unionTypeNode = (BLangUnionTypeNode) node;
+                for (BLangType member : unionTypeNode.memberTypeNodes) {
+                    validateInferredArrays(member, expr);
+                }
+                return;
+            default:
+                return;
+        }
+    }
+
+    private boolean validateInitializer(List<BLangExpression> sizes, int index, BLangListConstructorExpr expr) {
```

Comment:
Shall we change the parameter `sizes` into something more meaningful?

Reply:
Addressed in https://github.com/ballerina-platform/ballerina-lang/pull/34612/commits/dd4a64875eca77bcc52d1ceb139a8f36621b4cea